cdc/processor/pipeline/table_actor_test.go

```diff
@@ -247,26 +248,38 @@ func TestPollTickMessage(t *testing.T) {
 	}))
 }
 
-func TestPollStopMessage(t *testing.T) {
-	wg := sync.WaitGroup{}
-	wg.Add(1)
+func TestPollStopMessageAndSinkNodeStopReentrant(t *testing.T) {
 	tbl := tableActor{
-		state: TableStateStopped,
-		cancel: func() {
-			wg.Done()
-		},
+		state:     TableStateReplicating,
+		cancel:    func() {},
 		reportErr: func(err error) {},
 	}
 	tbl.sinkNode = &sinkNode{
 		state:          &tbl.state,
 		sinkV1:         mocksink.NewNormalMockSink(),
 		flowController: &mockFlowController{},
 	}
+
+	grs := runtime.NumGoroutine()
 	tbl.Poll(context.TODO(), []message.Message[pmessage.Message]{
 		message.StopMessage[pmessage.Message](),
 	})
-	wg.Wait()
-	require.Equal(t, stopped, tbl.stopped)
+	require.Eventually(t, func() bool {
+		return tbl.state.Load() == TableStateStopped
+	}, 10*time.Second, 10*time.Millisecond)
+	require.True(t, tbl.sinkStopped)
+	// Try to stop again, should not block and return immediately.
+	tbl.Poll(context.TODO(), []message.Message[pmessage.Message]{
+		message.StopMessage[pmessage.Message](),
+	})
+	tbl.Poll(context.TODO(), []message.Message[pmessage.Message]{
+		message.StopMessage[pmessage.Message](),
+	})
+	tbl.Poll(context.TODO(), []message.Message[pmessage.Message]{
+		message.StopMessage[pmessage.Message](),
+	})
+	// Check it immediately, should no more goroutine.
+	require.LessOrEqual(t, runtime.NumGoroutine(), grs+1)
```

Comment:
Do we have a more accurate test method? I think it will be an unstable test case if we just measure the goroutine count.

Reply:
Do you have any suggestions? Which scenario will be unstable?

Reply:
Please check out https://github.com/pingcap/tiflow/pull/7110/commits/120331cfdd776dd09bec9d416e2b0a89e497986b